[Docs] Add logs-and-relabeling-basics section to 'Flow by Example' tutorials (#6172)
grafana/agent: docs/sources/flow/tutorials/flow-by-example/logs-and-relabeling-basics/index.md

    ]
}

prometheus.relabel "example" {
    forward_to = [
        prometheus.remote_write.local_prom.receiver,
    ]

    rule {
        action       = "replace"
        target_label = "os"
        replacement  = constants.os
    }
}

prometheus.remote_write "local_prom" {
    endpoint {
        url = "http://localhost:9090/api/v1/write"
    }
}
```

We have now created the following pipeline:

<p align="center">
<img src="/media/docs/agent/diagram-flow-by-example-relabel-0.svg" alt="Diagram of pipeline that scrapes prometheus.exporter.unix, relabels the metrics, and remote_writes them." width="500" />
</p>

This pipeline has a `prometheus.relabel` component that has a single rule. This rule has the `replace` action, which will replace the value of the `os` label with a special value: `constants.os`. This value is a special constant that will be replaced with the os of the host the Agent is running on. You can see the other available constants in the [constants](https://grafana.com/docs/agent/<AGENT_VERSION>/flow/reference/stdlib/constants/) documentation page. In this example we have one rule block, but you can have as many as you want. Each rule block will be applied in order.

> If you run the Agent and navigate to [localhost:3000/explore](http://localhost:3000/explore), you should be able to see the `os` label on the metrics. Try querying for `node_context_switches_total` and look at the labels!

Author: Yeah, just some unnecessary weasel-words on my part :sweat_smile:

Relabeling uses the same rules as Prometheus. You can always refer to the [prometheus.relabel documentation](https://grafana.com/docs/agent/<AGENT_VERSION>/flow/reference/components/prometheus.relabel/#rule-block) for a full list of available options.

{{% admonition type="note" %}}
You can have multiple components forward to one `prometheus.relabel` component. This allows you to apply the same relabeling rules to multiple pipelines.
{{% /admonition %}}

{{% admonition type="warning" %}}

> There is a common "gotcha" when using metrics that start with `__` (double underscore). These labels are considered internal and will be dropped before relabeling rules from a `prometheus.relabel` component are applied. If you would like to keep or act on these kinds of labels, use a [discovery.relabel](https://grafana.com/docs/agent/<AGENT_VERSION>/flow/reference/components/discovery.relabel/) component.

Reviewer: I have been thinking about this one... and might have some better wording.

> There is a common "gotcha" when using metrics that start with `__` (double underscore). These labels are considered internal and will be dropped before relabeling rules from a `prometheus.relabel` component are applied. If you would like to keep or act on these kinds of labels, use a [discovery.relabel](https://grafana.com/docs/agent/<AGENT_VERSION>/flow/reference/components/discovery.relabel/) component.

Reviewer: ```suggestion
When metric labels start with `__` (double underscore), they are considered internal labels and are dropped before the relabeling rules from a `prometheus.relabel` component are applied. If you would like to keep or act on these labels, use a [discovery.relabel](https://grafana.com/docs/agent/<AGENT_VERSION>/flow/reference/components/discovery.relabel/) component.
```

Maybe this is better?

> There is a common "gotcha" when using metrics that start with `__` (double underscore). These labels are considered internal and will be dropped before relabeling rules from a `prometheus.relabel` component are applied. If you would like to keep or act on these kinds of labels, use a [discovery.relabel](https://grafana.com/docs/agent/<AGENT_VERSION>/flow/reference/components/discovery.relabel/) component.

Author: This caught a mistake on my part: that the issue is _labels_ that start with `__`, not metrics. However, it's not inherently an issue with the label starting with `__` (this is part of Prometheus) but rather an issue commonly faced by users trying to use or keep those labels, since they have already been dropped.

How does:

```suggestion
There is an issue commonly faced when relabeling and using labels that start with `__` (double underscore). These labels are considered internal and are dropped before relabeling rules from a `prometheus.relabel` component are applied. If you would like to keep or act on these kinds of labels, use a [discovery.relabel](https://grafana.com/docs/agent/<AGENT_VERSION>/flow/reference/components/discovery.relabel/) component.
```
sound?

> There is a common "gotcha" when using metrics that start with `__` (double underscore). These labels are considered internal and will be dropped before relabeling rules from a `prometheus.relabel` component are applied. If you would like to keep or act on these kinds of labels, use a [discovery.relabel](https://grafana.com/docs/agent/<AGENT_VERSION>/flow/reference/components/discovery.relabel/) component.

Reviewer: Yeah that works... and makes more sense now :-)

```river
local.file_match "tmplogs" {
    path_targets = [{"__path__" = "/tmp/flow-logs/*.log"}]
}

loki.source.file "local_files" {
    targets    = local.file_match.tmplogs.targets
    forward_to = [loki.write.local_loki.receiver]
}

loki.write "local_loki" {
    endpoint {
        url = "http://localhost:3100/loki/api/v1/push"
    }
}
```

The rough flow of this pipeline is:

<p align="center">
<img src="/media/docs/agent/diagram-flow-by-example-logs-0.svg" alt="Diagram of pipeline that collects logs from /tmp/flow-logs and writes them to a local loki instance" width="500" />
</p>

If you navigate to [localhost:3000/explore](http://localhost:3000/explore) and switch the Datasource to `Loki`, you should be able to query for `{filename="/tmp/flow-logs/log.log"}` and see the log line we created earlier. Try running:

```bash
echo "This is another log line!" >> /tmp/flow-logs/log.log
```

to add some more logs to the file. If you re-execute the query, you should see the new log lines.

<p align="center">
<img src="/media/docs/agent/screenshot-flow-by-example-log-lines.png" alt="Grafana Explore view of example log lines." />
</p>

> If you are curious how the Agent keeps track of where in a log file it is, you can take a look at `data-agent/loki.source.file.local_files/positions.yml`. If you delete this file, the Agent will start reading from the beginning of the file again, which is why it is desirable to keep the Agent's data directory in a persistent location.

Reviewer: ```suggestion
If you are curious how {{< param "PRODUCT_ROOT_NAME" >}} keeps track of where it is in a log file, you can look at `data-agent/loki.source.file.local_files/positions.yml`. 
If you delete this file, {{< param "PRODUCT_ROOT_NAME" >}} starts reading from the beginning of the file again, which is why keeping the {{< param "PRODUCT_ROOT_NAME" >}}'s data directory in a persistent location is desirable.
```

> ## Exercise for the Reader

Reviewer: ```suggestion
## Exercise
```

[loki.relabel]: https://grafana.com/docs/agent/<AGENT_VERSION>/flow/reference/components/loki.relabel/
[loki.process]: https://grafana.com/docs/agent/<AGENT_VERSION>/flow/reference/components/loki.process/

**Recommended reading**

- [loki.relabel][]
- [loki.process][]

### Add a Label to Logs

> This exercise will have two parts, building on the previous example. Let's start by adding an `os` label (just like the Prometheus example) to all of the logs we are collecting.

Reviewer: ```suggestion
This exercise will have two parts, building on the previous example. Let's start by adding an `os` label (just like the Prometheus example) to all of the logs we collect.
```

> Modify the following snippet to add label `os` with the value of the `os` constant.

Reviewer: ```suggestion
Modify the following snippet to add the label `os` with the value of the `os` constant.
```

```river
local.file_match "tmplogs" {
    path_targets = [{"__path__" = "/tmp/flow-logs/*.log"}]
}

loki.source.file "local_files" {
    targets    = local.file_match.tmplogs.targets
    forward_to = [loki.write.local_loki.receiver]
}

loki.write "local_loki" {
    endpoint {
        url = "http://localhost:3100/loki/api/v1/push"
    }
}
```

{{% admonition type="note" %}}

> The [loki.relabel](https://grafana.com/docs/agent/<AGENT_VERSION>/flow/reference/components/loki.relabel) component can be used to relabel and add labels, just like the [prometheus.relabel](https://grafana.com/docs/agent/<AGENT_VERSION>/flow/reference/components/prometheus.relabel) component.

Reviewer: ```suggestion
You can use the [loki.relabel](https://grafana.com/docs/agent/<AGENT_VERSION>/flow/reference/components/loki.relabel) component to relabel and add labels, just like you can with the [prometheus.relabel](https://grafana.com/docs/agent/<AGENT_VERSION>/flow/reference/components/prometheus.relabel) component.
```

{{% /admonition %}}

> Once you have your completed config, run the Agent and execute the following:

Reviewer: ```suggestion
Once you have your completed configuration, run {{< param "PRODUCT_ROOT_NAME" >}} and execute the following:
```

echo 'level=debug msg="DEBUG: This is a debug level log!"' >> /tmp/flow-logs/log.log
```

Navigate to [localhost:3000/explore](http://localhost:3000/explore) and switch the Datasource to `Loki`. Try querying for `{filename="/tmp/flow-logs/log.log"}` and see if you can find the new label!

Now that we have added new labels, we can also filter on them. Try querying for `{os!=""}`. You should only see the lines you added in the previous step.

{{< collapse title="Solution" >}}

```river
// Let's learn about relabeling and send logs to Loki!

local.file_match "tmplogs" {
    path_targets = [{"__path__" = "/tmp/flow-logs/*.log"}]
}

loki.source.file "local_files" {
    targets    = local.file_match.tmplogs.targets
    forward_to = [loki.relabel.add_static_label.receiver]
}

loki.relabel "add_static_label" {
    forward_to = [loki.write.local_loki.receiver]

    rule {
        target_label = "os"
        replacement  = constants.os
    }
}

loki.write "local_loki" {
    endpoint {
        url = "http://localhost:3100/loki/api/v1/push"
    }
}
```

{{< /collapse >}}

### Extract and add a Label from Logs

{{% admonition type="note" %}}

> This exercise is more challenging than the previous one. If you are having trouble, feel free to skip it and move to the next section, which will cover some of the concepts used here. You can always come back to this exercise later!

Reviewer: ```suggestion
This exercise is more challenging than the previous one. If you are having trouble, skip it and move to the next section, which will cover some of the concepts used here. You can always come back to this exercise later.
```

{{% /admonition %}}

> This exercise will build on the previous one, though it's a bit more involved. 

Reviewer: ```suggestion
This exercise will build on the previous one, though it's more involved. 
```

> Let's say we want to extract the `level` from the logs and add it as a label. As a starting point, take a look at [loki.process][]. This component allows you to perform processing on logs, including extracting values from log contents.

Reviewer: ```suggestion
Let's say we want to extract the `level` from the logs and add it as a label. As a starting point, look at [loki.process][].
This component allows you to perform processing on logs, including extracting values from log contents.
```

> Try modifying your config from the previous section to extract the `level` from the logs and add it as a label. If you need it, you can find a solution to the previous exercise at the end of the [previous section](#add-a-label-to-logs).

Reviewer: ```suggestion
Try modifying your configuration from the previous section to extract the `level` from the logs and add it as a label.
If needed, you can find a solution to the previous exercise at the end of the [previous section](#add-a-label-to-logs).
```

{{% admonition type="note" %}}
The `stage.logfmt` and `stage.labels` blocks for `loki.process` may be helpful.
{{% /admonition %}}

> Once you have your completed config, run the Agent and execute the following:

Reviewer: ```suggestion
Once you have your completed config, run {{< param "PRODUCT_ROOT_NAME" >}} and execute the following:
```

```bash
echo 'level=info msg="INFO: This is an info level log!"' >> /tmp/flow-logs/log.log
echo 'level=warn msg="WARN: This is a warn level log!"' >> /tmp/flow-logs/log.log
echo 'level=debug msg="DEBUG: This is a debug level log!"' >> /tmp/flow-logs/log.log
```

> Navigate to [localhost:3000/explore](http://localhost:3000/explore) and switch the Datasource to `Loki`. Try querying for `{level!=""}` to see the new labels in action!

Reviewer: ```suggestion
Navigate to [localhost:3000/explore](http://localhost:3000/explore) and switch the Datasource to `Loki`. Try querying for `{level!=""}` to see the new labels in action.
```

    targets    = local.file_match.tmplogs.targets
    forward_to = [loki.process.add_new_label.receiver]
}

loki.process "add_new_label" {
    // Extract the value of "level" from the log line and add it to the extracted map as "extracted_level"
    // You could also use "level" = "", which would extract the value of "level" and add it to the extracted map as "level"
    // but to make it explicit for this example, we will use a different name.
    //
    // The extracted map will be covered in more detail in the next section.
    stage.logfmt {
        mapping = {
            "extracted_level" = "level",
        }
    }

    // Add the value of "extracted_level" from the extracted map as a "level" label
    stage.labels {
        values = {
            "level" = "extracted_level",
        }
    }

    forward_to = [loki.relabel.add_static_label.receiver]
}

loki.relabel "add_static_label" {
    forward_to = [loki.write.local_loki.receiver]

    rule {
        target_label = "os"
        replacement  = constants.os
    }
}

loki.write "local_loki" {
    endpoint {
        url = "http://localhost:3100/loki/api/v1/push"
    }
}
```

{{< /collapse >}}

> ## Finishing up and next steps
> 
> Hopefully by now the concepts of components, attributes, and expressions are starting to make sense. You have also seen how to use some of the standard library components to collect metrics and logs. In the next section, you will learn more about how to use the `loki.process` component to extract values from logs and use them.

Reviewer: ```suggestion
## Finishing up and next steps

You have learned the concepts of components, attributes, and expressions. You have also seen how to use some standard library components to collect metrics and logs. In the next tutorial, you will learn more about how to use the `loki.process` component to extract values from logs and use them.

```
